src/core/server/plugins/plugins_service.ts

```diff
@@ -358,6 +338,48 @@ export class PluginsService implements CoreService<PluginsServiceSetup, PluginsS
     this.log.debug(`Discovered ${pluginEnableStatuses.size} plugins.`);
   }
 
+  /** Throws an error if the plugin's dependencies are invalid. */
+  private validatePluginDependencies(
+    plugin: PluginWrapper,
+    pluginEnableStatuses: Map<PluginName, { plugin: PluginWrapper; isEnabled: boolean }>
+  ) {
+    const { name, manifest, requiredBundles, requiredPlugins } = plugin;
+
+    // validate that `requiredBundles` ids point to a discovered plugin which `includesUiPlugin`
+    for (const requiredBundleId of requiredBundles) {
+      if (!pluginEnableStatuses.has(requiredBundleId)) {
+        throw new Error(
+          `Plugin bundle with id "${requiredBundleId}" is required by plugin "${name}" but it is missing.`
+        );
+      }
+
+      const requiredPlugin = pluginEnableStatuses.get(requiredBundleId)!.plugin;
+      if (!requiredPlugin.includesUiPlugin) {
+        throw new Error(
+          `Plugin bundle with id "${requiredBundleId}" is required by plugin "${name}" but it doesn't have a UI bundle.`
+        );
+      }
+
+      if (requiredPlugin.manifest.type !== plugin.manifest.type) {
+        throw new Error(
+          `Plugin bundle with id "${requiredBundleId}" is required by plugin "${name}" and expected to have "${manifest.type}" type, but its type is "${requiredPlugin.manifest.type}".`
+        );
+      }
+    }
+
+    // validate that OSS plugins do not have required dependencies on X-Pack plugins
+    if (plugin.source === 'oss') {
+      for (const id of [...requiredPlugins, ...requiredBundles]) {
+        const requiredPlugin = pluginEnableStatuses.get(id);
+        if (requiredPlugin && requiredPlugin.plugin.source === 'x-pack') {
+          throw new Error(
+            `X-Pack plugin or bundle with id "${id}" is required by OSS plugin "${name}", which is prohibited. Consider making this an optional dependency instead.`
```

Comment:
Should we be explicitly saying "OSS" in the error message here, since it isn't technically accurate anymore? (I know we still use it everywhere in the code, but thought I'd mention it in case it causes any confusion for 3rd party plugin devs). It seems to me we could probably omit it and the message would still make sense.

Reply:
That's a good question.

So, we do have lots of places in code, tests, and comments where we explicitly refer to "OSS". As far as I can tell though, we don't mention it to users directly and we don't mention it in log messages.

OTOH, we do still have an "OSS" build (even if we aren't actively producing those builds anymore), which is why we are only allowing _optional_ dependencies on X-Pack plugins.

I lean towards clarity here, and I think "OSS plugin" is the most correct term that we have. Otherwise we'd have to say something like:
_X-Pack plugin or bundle with id "foo" is required by non-X-Pack internal plugin "bar", which is prohibited._

Honestly, the only people who will see this message is people who are modifying Kibana's internal plugins, so I think this error message is fine. IMO, we can always change the error later if we want to scrub all references to "OSS" in our code.